### PR TITLE
Gym interface

### PR DESCRIPTION
**Pull request by manuel-delverme**
Only the simplified version works for now.

### CODE REVIEW

File: gym_gene_control/envs/grn_control_simple.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        return self.initial_state
+
+    def render(self, mode='human'):
+        raise NotImplemented
+
+    def close(self):
+        raise NotImplemented
+
+
+if __name__ == '__main__':
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:64]: there are nans at the beginning when i run it

File: gym_gene_control/envs/grn_control_simple.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        # TODO: return bad reward if there is no convergence
+        return x_T, reward, done, {}
+
+    def to_array(self, x):
+        expr_clean = jnp.stack(tuple([x[gene] for gene in range(self.sim.num_genes)])).swapaxes(0, 1)
+        return expr_clean
+
+    def reset(self):
+        if self.initial_state is None:
+            action = jnp.zeros(self.sim.num_genes)
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:53]: the actions are not those, but only the master regulators

**Review by ioneliabuzatu — CHANGES_REQUESTED** [review]

File: gym_gene_control/envs/grn_control_simple.py (hunk 2 above)
Comment (ioneliabuzatu) [reply gym_gene_control/envs/grn_control_simple.py:53]: actions = jnp.zeros(len(self.sim.layers[0]))

Comment (ioneliabuzatu) [reply gym_gene_control/envs/grn_control_simple.py:53]: actions = jnp.zeros(len(self.sim.layers[0]))

Comment (ioneliabuzatu) [reply gym_gene_control/envs/grn_control_simple.py:53]: updated: actions = jnp.ones(len(self.sim.layers[0]))

File: gym_gene_control/envs/grn_control_simple.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            num_genes=dataset.tot_genes,
+            num_cells_types=dataset.tot_cell_types,
+            simulation_num_steps=self.num_cells_to_sim,
+            interactions_filepath=dataset.interactions,
+            regulators_filepath=dataset.regulators,
+            noise_amplitude=self.noise_amplitude,
+        )
+        self.sim.build()
+        self.action_space = gym.spaces.Box(low=np.zeros(self.sim.num_genes), high=np.ones(self.sim.num_genes))
+        self.observation_space = gym.spaces.Box(low=np.zeros(self.sim.num_genes), high=np.ones(self.sim.num_genes))
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:31]: we are only controlling the master genes, not all of them. so len(self.sim.layers[0])

**Review by ioneliabuzatu — CHANGES_REQUESTED** [review]

File: gym_gene_control/envs/grn_control_simple.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        desired_concentration = self.initial_state[:, :, self.target_gene_type].mean(axis=0)  # TODO: worry about outliers, zeros and negatives
+        dist = jnp.linalg.norm(np.tile(desired_concentration.reshape(1, -1, 1), (1, 1, x_T.shape[2])) - x_T)
+
+        reward = -dist
+        done = False
+
+        # TODO: return bad reward if there is no convergence
+        return x_T, reward, done, {}
+
+    def to_array(self, x):
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:47]: could be dict_to_array to recall why we do to_array

File: gym_gene_control/envs/grn_control_simple.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        self.action_space = gym.spaces.Box(low=np.zeros(self.sim.num_genes), high=np.ones(self.sim.num_genes))
+        self.observation_space = gym.spaces.Box(low=np.zeros(self.sim.num_genes), high=np.ones(self.sim.num_genes))
+        self.initial_state = None
+
+    def step(self, action):
+        x_T = self.sim.run_one_rollout(action)
+        x_T = self.to_array(x_T)
+
+        desired_concentration = self.initial_state[:, :, self.target_gene_type].mean(axis=0)  # TODO: worry about outliers, zeros and negatives
+        dist = jnp.linalg.norm(np.tile(desired_concentration.reshape(1, -1, 1), (1, 1, x_T.shape[2])) - x_T)
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:39]: is this where the classifier should be instead?

File: gym_gene_control/envs/grn_control_simple.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    def close(self):
+        raise NotImplemented
+
+
+if __name__ == '__main__':
+    env = GRNControlSimpleEnv()
+    x = env.reset()
+    print(x)
+    for _ in range(100):
+        action = env.action_space.sample()
```
Comment (ioneliabuzatu) [inline gym_gene_control/envs/grn_control_simple.py:69]: right now action is of length 100 but we want only for the master genes

File: gym_gene_control/envs/grn_control_simple.py (hunk 3 above)
Comment (manuel-delverme) [reply gym_gene_control/envs/grn_control_simple.py:31]: Thank you.

**Review by ioneliabuzatu — CHANGES_REQUESTED** [review]

File: gym_gene_control/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+from gym_gene_control.envs.grn_control import GRNControlEnv
```
[the remark this answers is not in the document]
Comment (ioneliabuzatu) [reply gym_gene_control/__init__.py:1]: Okay sounds good.

File: gym_gene_control/envs/grn_control_simple.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        x_T = self.to_array(x_T)
+
+        desired_concentration = self.initial_state[:, :, self.target_gene_type].mean(axis=0)  # TODO: worry about outliers, zeros and negatives
+        dist = jnp.linalg.norm(np.tile(desired_concentration.reshape(1, -1, 1), (1, 1, x_T.shape[2])) - x_T)
+
+        reward = -dist
+        done = False
+
+        # TODO: return bad reward if there is no convergence
+        return x_T, reward, done, {}
```
[the remark this answers is not in the document]
Comment (manuel-delverme) [reply gym_gene_control/envs/grn_control_simple.py:45]: What would you like to be included as info?

**Pull request merged**